### PR TITLE
THREADING: Make CRYPTO_MUTEX and CRYPTO_CONDVAR typesafe

### PR DESCRIPTION
**Pull request by hlandau**
There was really no need for this to be void and it made bugs very easy
to introduce accidentally, especially given that the free functions
needed to be passed a pointer to the pointer.

Also fix some bugs in the QUIC code detected immediately by this change.

### CODE REVIEW

**Review by t8m — APPROVED** [review]
Approved if CI passes

**Comment by t8m** [comment]
Should this go into 3.2 as well?

**Comment by hlandau** [comment]
> Should this go into 3.2 as well?

@mattcaswell Your call IMO &mdash; is this within stable release guidelines? The bugfixes certainly should of course.

**Comment by t8m** [comment]
If the change of the type (which is strictly saying a refactor) is accompanied with a bug fix, it should IMO go in 3.2.

**Comment by hlandau** [comment]
> If the change of the type (which is strictly saying a refactor) is accompanied with a bug fix, it should IMO go in 3.2.

I could just fix the bugs detected by it in a separate PR. So I'm not sure we can really argue that.

**Comment by t8m** [comment]
> > If the change of the type (which is strictly saying a refactor) is accompanied with a bug fix, it should IMO go in 3.2.
> 
> I could just fix the bugs detected by it in a separate PR. So I'm not sure we can really argue that.

I know you can just fix the bugs, but potentially further ones could be introduced and the change of the type should not have any negative impact. There is IMO no requirement that the bug fix must be absolutely minimal and we backported similar changes with no negative impact along with related bug fixes in past.

**Comment by mattcaswell** [comment]
> @mattcaswell Your call IMO — is this within stable release guidelines? The bugfixes certainly should of course.

Not really my call anymore than anyone else's. But I'll give my opinion.

This PR is clearly a bug fix that is eligible for backport to 3.2. In general broader "cleanup" is not eligible except where it is an intrinsic part of the bug fix. IMO, in this case it really is. The "cleanup" up part is a really small and non-risky change which makes perfect sense in the context of this bug fix. So I think it is ok.

**Review by mattcaswell — APPROVED** [review]

**Pull request closed without merging**
